pennylane/transforms/batch_input.py

```diff
@@ -73,25 +72,39 @@ def circuit(inputs, weights):
     array([0.46230079, 0.73971315, 0.95666004, 0.5355225 , 0.66180948,
             0.44519553, 0.93874261, 0.9483197 , 0.78737918, 0.90866411])>
     """
-    parameters = tape.get_parameters(trainable_only=False)
-
     argnum = tuple(argnum) if isinstance(argnum, (list, tuple)) else (int(argnum),)
 
-    non_trainable, trainable = [], []
-    for idx, param in enumerate(parameters):
-        if idx in argnum:
-            non_trainable.append(param)
-        else:
-            trainable.append(param)
+    trainable_params = tape.get_parameters()
+    all_parameters = sum(
+        (op.parameters if op.parameters != [] else [None] for op in tape.operations), []
+    )
+    argnum_params = [all_parameters[i] for i in argnum]
 
-    if len(np.unique([qml.math.shape(x)[0] for x in non_trainable])) != 1:
+    if any(
+        any(param is train_param for train_param in trainable_params) for param in argnum_params
+    ):
```

Comment:
This seems to iterate quite a lot over the trainable parameters, which might be a big sequence of parameters, right?
Is there a more efficient way? Maybe something like
```suggestion
    if any(qml.math.requires_grad(p) for p in argnum_params):
```

Reply:
Alternatively, we could check for overlap between `tape.trainable_params` and `argnum` ? :)

Reply:
Here I wanted to use: `param in trainable_params for param in argnum_params`, but this is ill-defined when `param` and `trainable_params` contain multiple dimension tensors.

I didn't know of `qml.math.requires_grad`! This is definitely better.